### PR TITLE
Remove hardcoded shortcuts from /scene and instead use the input action system to allow them to be customised.

### LINKED ISSUE #1532: Allow hard-coded keyboard shortcuts in GUI controls to be customised
### Describe the problem or limitation you are having in your project:
In the Godot engine, a number of GUI controls have keyboard shotcuts 'built-in' to the C++ code. For example, in Text Edit, Copy and Paste are hard-coded to `Ctrl + C` and `Ctrl + V` respectively. This can cause issues and confusion when trying to rebind input that uses these hard-coded shortcuts.

Often these shortcuts give the impression that they can be rebound... Take a look at the Script Text Editor shortcuts in Editor Settings.
image
However, even if you change this binding,  `Ctrl + C` and `Ctrl + V` still both work for copy and paste, since they are built in to the TextEdit code, which ScriptTextEditor uses under the hood.

Related issues:
https://github.com/godotengine/godot/issues/29490
https://github.com/godotengine/godot/issues/42139
https://github.com/godotengine/godot/issues/12164
https://github.com/godotengine/godot/issues/17927
https://github.com/godotengine/godot/issues/26854
https://github.com/godotengine/godot/issues/31739 (this is loosely related - but the comment @akien-mga makes at the end is relevant - _"A recent example: in 28561 I added a feature to convert clipboard indentation on paste, but I can't have it overwrite Ctrl+V because that's hardcoded in TextEdit. The new API should allow overriding TextEdit's input management in CodeEdit and EditorCodeEdit."_
https://github.com/godotengine/godot/issues/20065 - Ideally this proposals would help make the built-in editor shortcuts more obvious.

### **Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
A method to be able to rebind these shortcuts which is completely separate to EditorSettings to avoid breaching https://github.com/godotengine/godot/issues/29730.

### **Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

**PLEASE NOTE: The implementation has changed a lot since this was first written. I am now using the InputAction system. Please read the latest comments on this post for a clearer picture of how it will work.**

I have a test branch working already. Essentially, each GUI Control node would keep its own mini version of EditorSettings shortcuts which only applies to itself. Here is a mockup:

**control.cpp**
```cpp
// *** Header *** 
HashMap<StringName, Ref<InputEvent>> built_in_shortcuts;

// *** cpp ***
void Control::_set_built_in_shortcut(const StringName &p_name, const Ref<InputEvent> &p_shortcut); // Protected
Ref<InputEvent> Control::get_builtin_shortcut(const StringName &p_name); // Public, or Protected if option 1) is chosen below.
void Control::override_builtin_shortcut(const StringName &p_name, const Ref<Shortcut> &p_shortcut); // Public. Checks if p_name is a registered shortcut first. Not needed if option 1) is chosen below.
```

From here, I think there are 2 options.
1. Controls which derive from Control define their own setters and getters for their own shortcuts. These would call `_set_built_in_shortcut` and `get_builtin_shortcut` internally. This would remove the need for `override_builtin_shortcut` In this example text edit has quite a few shortcuts so there would be a number of methods:
```cpp
TextEdit::set_shortcut_copy(Ref<InputEvent> p_ev) {
	_set_built_in_shortcut("copy", p_ev);
}

// And other methods...
TextEdit::get_shortcut_copy();
TextEdit::set_shortcut_paste(...);
TextEdit::get_shortcut_paste();
```
2. Use a string based implementation. Instead of methods on the derived child, use the methods in `Control` to set the shortcuts. This would rely on strings (fragile) but would not require methods on the GUI control for every shortcut.

Either 1) or 2) would work... Built-in shortcuts don't change very often and C++ user's would not be adding them on the fly (like they can with EditorSettings), so this points to 1) being the better option as it does not rely on fragile strings and would be a more consistent and easy to understand API.

Ok, moving on... For brevity, I will use option 1) for the following examples.

**Usage in nodes which derive from control, e.g. TextEdit**
```cpp
// Inside the Constructor, set the default key binding.
Ref<InputEventKey> ie;
ie.instance();
ie->set_control(true);
ie->set_keycode(KEY_C);
set_shortcut_copy(ie);

// Inside _gui_input, or wherever the shortcut needs to be checked:
if (get_shortcut_copy()->shortcut_match(p_ev)) {
	// < logic for shortcut >
}

// For comparison, the current implementation is essentially:
Ref<InputEventKey> k = p_ev;
if (k->get_keycode() == KEY_C && k->get_control() ) { // Oh dear... the shortcut can't be changed!
	// < logic for shortcut >
}
```

**Usage when you want to override the shortcut... E.g. In ScriptTextEditor**
```cpp
// Inside ScriptTextEditor Constructor
code_editor->get_text_editor()->set_shortcut_copy(ED_GET_SHORTCUT("script_text_editor/copy", TTR("Copy"));
// Or, if you define the shortcut at the same time:
code_editor->get_text_editor()->set_shortcut_copy(ED_SHORTCUT("script_text_editor/copy", TTR("Copy"), KEY_MASK_CMD | KEY_C);
```

Now you can change the TextEdit shortcuts via an Editor Shortcut if you like, or you can just set a custom shortcut without using EditorSettings.

Note that ED_SHORTCUT is not being checked directly in the TextEdit, so if you make a change to EditorSettings it will no be used until the editor is restarted. I believe you can make the shortcuts update after a change if you handled `NOTIFICATION_EDITOR_SETTINGS_CHANGED` in `notification()` and just called `set_shortcut_...(ED_GET_SHORTCUT(...))` again. Assuming the notification is called on shortcut change - that might have to be added, but this is a small detail overall.

### LINKED ISSUE #12164: Custom keyboard shortcuts don't override default behavior
**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Godot 2.1.4
Godot 3.0 (built from master)

**Issue description:**
When using custom shortcuts, default behavior is executed before shortcut-behavior is executed.

**Steps to reproduce:**
Override keyboard shortcuts for 'Script Editor / Next Script' with Ctrl+PageDown
Override keyboard shortcuts for 'Script Editor / Previous Script' with Ctrl+PageUp

Use shortcuts in script editor.
When pressing Ctrl+PageUp see that cursor jumps to the first line of the file before going to previous script.
When pressing Ctrl+PageDown see that cursor jumps to the last line of the file before going to next script.

### LINKED ISSUE #17927: Can't reassign Cut command (Ctrl+X always cuts)
**Godot version:**
 3.0.2

**OS/device including version:**
Windows 10 64bit

**Issue description:**
I like to use Control+X for redo but I can't seem to override the shortcut for it. In the Script Editor > Edit menu, it shows up correctly and pressing it there works, but pressing Control+X still cuts the line and doesn't redo.

Even just removing the Control+X assignment doesn't work, it still cuts.

Tested on "blank" install (removed user prefs, let it remake it's own, then removed cut command) as well.

**Steps to reproduce:**
These are the preferences that were changed in the editor-settings-3.tres file on a clean install after removing the cut command assignment from the editor.

```
shortcuts = [ "script_text_editor/capitalize", SubResource( 1 ), "script_text_editor/convert_to_lowercase", SubResource( 2 ), "script_text_editor/convert_to_uppercase", SubResource( 3 ), "script_text_editor/cut", null, "script_text_editor/redo", SubResource( 4 ) ]
```

### LINKED ISSUE #26854: CTRL + SHIFT + UP/DOWN remapping not working in Script Text Editor
**Godot version:**
3.1 RC1

**OS/device including version:**
Windows 10 x64

**Issue description:**
I tried to remap the keyboard shortcuts "Move Up" and "Move Down" to become similar to the "Sublime Text" editor, which would respectively be `Ctrl+Shift+Up` and `Ctrl+Shift+Down`, but these combinations are not working.

**Steps to reproduce:**

Editor -> Editor Setting -> Shortcuts

### LINKED ISSUE #29490: Some shortcuts are hard-coded in TextEdit
**Godot version:**
3.1.1

**Issue description:**
I remember seeing similar issue (maybe), but can't find it now. Anyways, I was looking how #27164 could be implemented and thought that it would make sense if MenuButton shortcuts allowed for echo events. Not all of course, but for stuff like moving or duplicating lines that would be useful.

Then I noticed that some actions, like undo or paste, already support that. Tracking down the code, these actions are hard-coded in TextEdit. There are few problems with that. We can already remap actions like paste to another shortcut and it changes in script editor menu (but not the right-click menu of TextEdit, which is part of script editor). But doing this, we end up with 2 shortcuts for the same action, with one of them working as intended and the new one not supporting continuous actions.

So these actions should somehow be changed to proper shortcuts and also shortcuts should support optional echo events, so everything works the same, but is cleaner and allows improving other actions with continuous activation.

### LINKED ISSUE #42139: Can't reassign "ctrl +z" and "ctrl + y" keybinds on a Swiss QWERTZ keyboard
**Godot version:**
Godot Engine v3.2.3.stable.mono.official

**OS/device including version:**
Windows 10 Education N
Version 1903
Swiss Keyboard (the "Y" and "Z" keys are inverted)

**Issue description:**
Changed the keyboard shortcuts for "undo" and "redo". They accept the new keybinds, but the old ones ("ctrl + z" and "ctrl + y") remain active too.
Assigning "ctrl + z" and "ctrl + y" to different commands does not help, they remain bound to the "undo" and "redo" commands.

Note: The reason I want to reassign those keybinds is to invert them. the "Z" key is very far from the "ctrl" key on my swiss keyboard, making using the "undo" command painful in the long run.

**Steps to reproduce:**
1) Try to assign "ctrl + Z" and "ctrl + Y" keybinds to different commands
2) that's it

**Minimal reproduction project:**
[HelloGodot.zip](https://github.com/godotengine/godot/files/5240992/HelloGodot.zip)

### PR DESCRIPTION
**Pull request by EricEzaM**
* Made shortcuts which were previously hardcoded customisable. (this is just the backend implementation - the UI to do so will come in a later PR)
* Shortcuts that fall into this category are things like cut, copy paste, etc on TextEdit. In the code, these were defined in `gui_input` with things like 
`if (k->is_pressed() && k->get_keycode() == KEY_C && k->get_command() == true){ /* copy */}`
* Cleaned up logic... it is much more readable now. The below code example handles 1) moving the cursor left 1 char, 2) moving the cursor left by 1 word 3) moving the cursor to the start of the line (on Apple only) and finally 4) doing the previous 3 with selection (shift) pressed.

https://github.com/godotengine/godot/blob/07112b92a85de9b2134f3208149b295f871d3642/scene/gui/text_edit.cpp#L2876-L2945

Quite hard to follow in my opinion. This logic now looks like this, except the following logic can use any key combination, not just the left arrow key.
```cpp
// In gui_input:
if (k->is_action("ui_text_caret_left")) {
	_move_cursor_left(shift_pressed, false);
}
if (k->is_action("ui_text_caret_word_left")) {
	_move_cursor_left(shift_pressed, true);
}
// …
if (k->is_action("ui_text_caret_line_start")) {
	_move_cursor_to_line_start(shift_pressed);
}
// Other methods in the class:
void TextEdit::_move_cursor_left(bool p_select, bool p_move_by_word);
void TextEdit::_move_cursor_to_line_start(bool p_select);
```

The keys used to perform these actions are defined in InputMap:
```cpp
inputs = List<Ref<InputEvent>>();
inputs.push_back(InputEventKey::create_reference(KEY_HOME));
map.insert("ui_text_caret_line_start", inputs);

inputs = List<Ref<InputEvent>>();
inputs.push_back(InputEventKey::create_reference(KEY_A, KEY_MASK_CTRL));
inputs.push_back(InputEventKey::create_reference(KEY_LEFT, KEY_MASK_CMD));
map.insert("ui_text_caret_line_start.OSX", inputs);
```

**Added Utility Method**
Added static methods `::create_reference` on InputEventKey and JoypadButton which is a one-line way to do:
```
Ref<InputEventJoypadButton> ie;
ie.instance();
ie->set_button_index(p_btn_index);
```
The above code is used in many places with Buttons and Keys, so moving it to a utility one-liner saves A LOT of lines of code, at least a couple hundred lines.

**Unified definitions of the same shortcuts**
Moved builtin input actions away from being defined twice in ProjectSettings and EditorSettings to only being defined once in InputMap. Project/Editor Settings instead just get the builtin actions and loop through them to add them to their settings list. Reduces code duplication.

Closes #17927
Closes #29490
Closes #12164
Closes #42139
Closes #26854
Closes godotengine/godot-proposals#1532

This is the fourth PR of a larger series based on #42600.

### CODE REVIEW

**Comment by EricEzaM** [comment]
Requesting review from @Paulb23 as we had quite the dialog about this PR in the associated proposal thread.

**Comment by akien-mga** [comment]
For the reference, here's the list of default actions in the InputMap as of this PR. I know #42600 might refine this further but this might still be worth a discussion on whether all of those should really be configurable (like caret up/right bound to <kbd>Up</kbd>/<kbd>Right</kbd>), or if they should be renamed/better sorted to make it more obvious what is being affected by those.

My main concern seeing this list is that it's huge for something that most users won't even want to configure. On the other hand the same can likely be said of some of the original `ui_` actions prior to this. (Though many users tend to reuse `ui_<direction>` and `ui_accept` for their game logic on simple projects.)

So IMO it's most important to ensure that the InputMap editor in the Project Settings will be as user-friendly as possible, even more so than the Editor Settings' shortcut configuration tab which is more advanced / less critical to each project. Again, this might be handled in #42600 already but discussion there seemed to focus on the Editor Settings. I'll give it a try to see.

User-defined actions and advanced UI configuration should probably be separated so that users can focus on their game actions first and foremost, and cater to fine tuning of Control actions only when they needed it (very rare cases IMO).

Screenshot_20201123_115805
Screenshot_20201123_115834
Screenshot_20201123_115849
Screenshot_20201123_115905
Screenshot_20201123_115936

**Comment by EricEzaM** [comment]
@akien-mga There should also be a full list at input_map.cpp @ line 267

Here I put the builtin action display names, for cases where they need to be displayed in a popup menu (like cut, copy paste, etc).
```
static const _BuiltinActionDisplayName _builtin_action_display_names[] = {
	/* clang-format off */
	{ "ui_accept", 									"Accept" },
	{ "ui_select", 									"Select" },
	{ "ui_cancel", 									"Cancel" },
	{ "ui_focus_next", 								"Focus Next" },
	{ "ui_focus_prev", 								"Focus Prev" },
	{ "ui_left", 									"Left" },
	{ "ui_right", 									"Right" },
	{ "ui_up", 										"Up" },
	{ "ui_down", 									"Down" },
	{ "ui_page_up",									"Page Up" },
	{ "ui_page_down", 								"Page Down" },
	{ "ui_home", 									"Home" },
	{ "ui_end", 									"End" },
	{ "ui_cut", 									"Cut" },
	{ "ui_copy", 									"Copy" },
	{ "ui_paste", 									"Paste" },
	{ "ui_undo", 									"Undo" },
	{ "ui_redo", 									"Redo" },
	{ "ui_text_completion_query", 					"Completion Query" },
	{ "ui_text_newline", 							"New Line" },
	{ "ui_text_newline_blank",						"New Blank Line" },
	{ "ui_text_newline_above", 						"New Line Above" },
	{ "ui_text_indent", 							"Indent" },
	{ "ui_text_dedent", 							"Dedent" },
	{ "ui_text_backspace", 							"Backspace" },
	{ "ui_text_backspace_word", 					"Backspace Word" },
	{ "ui_text_backspace_word.OSX", 				"Backspace Word" },
	{ "ui_text_backspace_all_to_left", 				"Backspace all to Left" },
	{ "ui_text_backspace_all_to_left.OSX", 			"Backspace all to Left" },
	{ "ui_text_delete", 							"Delete" },
	{ "ui_text_delete_word", 						"Delete Word" },
	{ "ui_text_delete_word.OSX", 					"Delete Word" },
	{ "ui_text_delete_all_to_right", 				"Delete all to Right" },
	{ "ui_text_delete_all_to_right.OSX", 			"Delete all to Right" },
	{ "ui_text_caret_left", 						"Caret Left" },
	{ "ui_text_caret_word_left", 					"Caret Word Left" },
	{ "ui_text_caret_word_left.OSX", 				"Caret Word Left" },
	{ "ui_text_caret_right", 						"Caret Right" },
	{ "ui_text_caret_word_right", 					"Caret Word Right" },
	{ "ui_text_caret_word_right.OSX", 				"Caret Word Right" },
	{ "ui_text_caret_up", 							"Caret Up" },
	{ "ui_text_caret_down", 						"Caret Down" },
	{ "ui_text_caret_line_start", 					"Caret Line Start" },
	{ "ui_text_caret_line_start.OSX", 				"Caret Line Start" },
	{ "ui_text_caret_line_end", 					"Caret Line End" },
	{ "ui_text_caret_line_end.OSX",					"Caret Line End" },
	{ "ui_text_caret_page_up", 						"Caret Page Up" },
	{ "ui_text_caret_page_down", 					"Caret Page Down" },
	{ "ui_text_caret_document_start", 				"Caret Document Start" },
	{ "ui_text_caret_document_start.OSX",			"Caret Document Start" },
	{ "ui_text_caret_document_end", 				"Caret Document End" },
	{ "ui_text_caret_document_end.OSX", 			"Caret Document End" },
	{ "ui_text_scroll_up", 							"Scroll Up" },
	{ "ui_text_scroll_up.OSX", 						"Scroll Up" },
	{ "ui_text_scroll_down", 						"Scroll Down" },
	{ "ui_text_scroll_down.OSX", 					"Scroll Down" },
	{ "ui_text_select_all", 						"Select All" },
	{ "ui_text_toggle_insert_mode", 				"Toggle Insert Mode" },
	{ "ui_graph_duplicate", 						"Duplicate Nodes" },
	{ "ui_graph_delete", 							"Delete Nodes" },
	{ "ui_filedialog_up_one_level", 				"Go Up One Level" },
	{ "ui_filedialog_refresh", 						"Refresh" },
	{ "ui_filedialog_show_hidden", 					"Show Hidden" },
	{ "",											""}
	/* clang-format on */
};
```

**Comment by EricEzaM** [comment]
Hmm those descriptions will also need a `TTRC()` I think

**Review by Paulb23 — COMMENTED** [review]
Code looks fine to me, I think this level of configuration is okay, mainly comes down to having a suitable UX, which is easier said then done. The other point is needing someway to document or as perhaps as part of the UX, that appending `.OSX` overrides it for apple keyboards. Though similar to Aiken suggestion in #42600, it might be worth doing the UX changes as a separate proposal / PR so we don't get bogged down here.

Needs a hefty rebase after #41100, which also adds a new shortcut `ui_swap_input_direction = KEY_QUOTELEFT` to `LineEdit` and `TextEdit`.

Lastly, found #22927 which this should close.

**Comment by EricEzaM** [comment]
Rebased.

I added the `ui_swap_input_direction` action mapping and created a method for it in LineEdit and TextEdit.

**Review by Paulb23 — DISMISSED** [review]
Got round to testing this locally, seem like we need to fix or workaround #18793 first otherwise we can't determine the difference between `Tab` and `Ctrl+Tab` etc. 

Also spotted a couple more editor shortcuts need updating [here.](https://github.com/godotengine/godot/blob/a132e89b6aca09bd0f0871dca91c1aa648dbad6a/editor/plugins/text_editor.cpp#L587-L608)

**Comment by EricEzaM** [comment]
> Got round to testing this locally, seem like we need to fix or workaround #18793 first otherwise we can't determine the difference between `Tab` and `Ctrl+Tab` etc.

This is caused by #44180

**Comment by Paulb23** [comment]
While testing this with #44355, it looks like some changes got lost with the #41100 rebase. `_move_cursor_left` moves the caret right by word and `_move_cursor_right` has not been updated to use `TextServer`.

**Comment by EricEzaM** [comment]
Fixed above issues @Paulb23.

I tested every single shortcut in TextEdit and LineEdit, and they are working correctly. Yes, #18793 broke it, however if you order the shortcuts correctly, it will work as intended. For example, take the `ui_text_newline_*` actions...

```
ui_text_newline_above = Control + Shift + Enter
ui_text_newline_blank = Control + Enter
ui_text_newline = Enter
```

If they are placed in order of specificity (as above), then it works ok. Same with Tab and Control+Tab. However, since these shortcuts are customisable, this is **not** an acceptable fix - it is only acceptable for testing the default options. #44355 or a similar fix to this issue will be required.

Note I also added a new `ui_` action, `ui_text_completion_accept`, because previously it was only using `ui_accept`, but we want to use `KEY_TAB` to accept a completion too, but `KEY_TAB` shouldn't be in `ui_accept`.

**Comment by akien-mga** [comment]
Will need a rebase after recent changes to TextEdit.

**Comment by EricEzaM** [comment]
Rebased

**Comment by EricEzaM** [comment]
I have added the use of `exact_match` and tested all the actions in the script editor, and it is working well!

File: core/input/input_map.cpp [hunk 1]
```diff
@@ -260,84 +260,444 @@ void InputMap::load_from_project_settings() {
...
+	{ "ui_text_dedent", 							TTRC("Dedent") },
+	{ "ui_text_backspace", 							TTRC("Backspace") },
+	{ "ui_text_backspace_word", 					TTRC("Backspace Word") },
+	{ "ui_text_backspace_word.OSX", 				TTRC("Backspace Word") },
+	{ "ui_text_backspace_all_to_left", 				TTRC("Backspace all to Left") },
+	{ "ui_text_backspace_all_to_left.OSX", 			TTRC("Backspace all to Left") },
+	{ "ui_text_delete", 							TTRC("Delete") },
+	{ "ui_text_delete_word", 						TTRC("Delete Word") },
+	{ "ui_text_delete_word.OSX", 					TTRC("Delete Word") },
+	{ "ui_text_delete_all_to_right", 				TTRC("Delete all to Right") },
```
Comment (akien-mga) [inline core/input/input_map.cpp:302]: Aligning code with tabs is never a good option :)
Screenshot_20210216_100307

Either use spaces to ensure exact alignment, or remove the alignment.

**Comment by EricEzaM** [comment]
All done, changed to spaces :)

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Needs another rebase after #46127 (then can be merged once CI passes).

**Comment by akien-mga** [comment]
Thanks!

**Comment by EricEzaM** [comment]
It's in! How exciting!

I'm really happy with how this simplifies the logic in gui input for text and line edit

**Comment by christinoleo** [comment]
@EricEzaM Thanks for the work here.
I have a question to you. It seems that due to commits 074f535 and 49714b0 (maybe there are more, but this is as far as I got), by default the autocompletion in the editor accepts Space, Enter and KP_Enter. The space one is causing (at least me) to always when editing a gdscript to autocomplete AABB when typing "<variable> = ". The last space autocompletes to "<variable> =AABB" instead of leaving a space. Of course I can just remove space from the ui_accept shortcuts, but it took me some days to figure this out since godot 3.2 does not have this issue (at least by default). In your opinion, do you think it's worth not having the spacebar as a default ui_accept or is there some other code change which is actually causing this weird autocomplete do be done? I intended to create a new issue ticket but I wish to get your opinion first.. 
aaaa
Thanks!

**Comment by EricEzaM** [comment]
@christinoleo Thanks for pointing this out! Perhaps a new action needs to be created, specifically for accepting auto-complete options, which only includes tab, enter and kp_enter.

**Comment by Calinou** [comment]
> @christinoleo Thanks for pointing this out! Perhaps a new action needs to be created, specifically for accepting auto-complete options, which only includes tab, enter and kp_enter.

How should we implement this for use in the editor? Having to edit project-specific settings to tweak autocompletion to your liking sounds bad. Personally, I'd prefer to have only <kbd>Tab</kbd> accept suggestions, but I understand most people want <kbd>Enter</kbd> to accept suggestions as well.

**Comment by EricEzaM** [comment]
@Calinou never mind actually - it is a very easy fix. For some reason I check for both `ui_text_completion_accept` and `ui_accept` in the code... It should really only be the former.

https://github.com/godotengine/godot/blob/ed2f51b15fee22a35d2a975fd77a70fc291bd8de/scene/gui/text_edit.cpp#L3262-L3266

**Comment by YuriSizov** [comment]
Which shortcut have you modified, and what exactly are you trying to do to test it?

**Comment by Thane5** [comment]
I set "ui_cut" to ctrl+shift+x
image

For testing, i created a new project. Created a "control" scene, added a button. 

If i select it and press ctrl+shift+x, nothing happens. If i press ctrl+x, the object gets cut.

**Comment by akien-mga** [comment]
Works fine for me on Linux with 4.0 RC 3:

image

Selecting text in the script editor and pressing Ctrl+Shift+X cuts it. Pressing Ctrl+X writes "x".

**Comment by Thane5** [comment]
In the text editor, it works for me too, but not in the 2D or 3D editor.

**Comment by Thane5** [comment]
Oh, i just realised that there is a seperate "cut" command for the scene tree. I did not find it at first since i use the german version of the editor and unlike "ui_cut", the "scene tree cut" has already been translated into german.

However i find it a bit weird that "ui_undo" and "ui_redo" can affect the scene tree as expected, while "ui_cut" is specific to the text editor.

Anyways, at least my issue is solved now.
